### ISSUE TITLE
GimKit kicks users "Cheating Detected"

### ISSUE THREAD

**Issue by FloppyT** [ecc521/gimkit-bot#8]
Now that Gimkit placed this filter - almost all of the hacks on Github have failed to work. This filter checks your progress over the questions and your speed.

**Comment by FloppyT** [comment]
I have info that Gimkit will allow host reconnections

**Comment by ObviousNoise** [comment]
do you know what the threshold is, like how fast can we answer questions and still fly under the radar?

**Comment by ecc521** [comment]
I haven't been putting much effort into maintaining this project, so no, I
am not aware of what the limit is, and even if I was, it can always change.
You can iteratively increase the delays to figure out what the limit is,
and (hopefully) submit a PR incorporating the increased delays.

**Comment by ecc521** [comment]
I'm going to take another look at this and see if adjusting the delays works. As stated previously, any help on this project would be extremely useful.

**Comment by ecc521** [comment]
Simply pasting the bot code into the console and doing nothing triggers this message. It must be a targeted detection system that is easy to bypass.

**Comment by ecc521** [comment]
We should be able to use an IIFE to prevent this issue.

**Comment by FloppyT** [comment]
<img width="240" alt="Screen Shot 2020-05-17 at 8 08 27 AM" src="https://user-images.githubusercontent.com/34033288/82144933-99c85680-9815-11ea-863b-d86f219042a0.png">
They track "Exploit" from SnowLord's script, and "clickElement" from yours.

If you want to solve the issue wrap the entire code inside of a 
```js
(function() {
//Insertcode
// All variables will stay local
})()
```

**Comment by FloppyT** [comment]
I believe `zr` is kind of like sendPacket and `Ht.redboat` is the redboat packet type which will be sent to the server so that the server can kick. This hack checker loop occurs every second, on a setInterval loop.
The `.post` you see above tracks this data:
```js
{
  r: "clickElement", // script type
  n: "name", // will appear as "" if not ingame
  s: "classic" // type of gamemode
}
```

**Comment by FloppyT** [comment]
As for the rate limiting thing inside of the server. Just click one answer wrong every 10 times or so, that should solve it.

**Comment by ecc521** [comment]
The IIFE fix had already been implemented into a separate branch, around the time of my comment. Those changes are being merged now.

**Comment by ecc521** [comment]
It looks like GimKit has changed some design aspects, as I am encountering errors in the script. I'll need to do a few more fixes before I can figure out what is going on with the rate limiting.

**Comment by FloppyT** [comment]
Sounds great, I'm going to release a PR to neaten up the code and fix some minor bugs later today aswell

**Comment by ecc521** [comment]
I am no longer receiving the cheating detected message, however am receiving the "Answering too quickly". Luckily, there is a big difference in how the two are handled by GimKit.

**Comment by FloppyT** [comment]
I think there might be a way, let me just double check something

**Comment by FloppyT** [comment]
Something like this would work.

```js
Object.defineProperty(window,"clickElement", {
    get: ()=> false,
    set: function(selector) {
        document.querySelector(selector).click();
        this.value = false; // I don't know if this is necessary
    }
})
```

When Gimkit checks if it is set by typing something like `window["clickElement"]` they will get false - therefore unset, but you could call the function by actually setting it. Weird idea, but it would work.

**Comment by ecc521** [comment]
Fancy idea, but I'd rather not. GimKit could change their code to detect
this, as they did previously. An IIFE is much better.

**Comment by ecc521** [comment]
UPDATE: GimKit has started dropping perma-bans on IP Addresses, including mine, which has made development work quite difficult. I do not expect this to last very long however, as many schools utilize shared devices, where a single ban could easily impact large numbers of users and cause havoc.

**Comment by ObviousNoise** [comment]
Reset your IP, I can walk you throw that if you dont know how

**Comment by ecc521** [comment]
I'm probably going to end up using a VPN. Getting public IPs reassigned is ISP dependent. 
What were you going to say?

**Comment by FloppyT** [comment]
Well, they are most likely dropping these bans because they don't want to deal with this annoying stuff, they are working on a new thing called Ink apparently - so they are probably busy with that. 

The SS I shared with you is how they got your IP...

**Comment by ecc521** [comment]
The SS? Screen Shot? That is a GitHub URL.

**Comment by FloppyT** [comment]
Above I thought I sent a screen shot of the code where they log your IP... Ill resend.
Also, what does it show when you got IP Banned? Could you show a picture? or does it just not connect?

**Comment by ecc521** [comment]
<img width="428" alt="Screen Shot 2020-05-17 at 5 40 40 PM" src="https://user-images.githubusercontent.com/42424091/82160751-88a73600-9865-11ea-89cc-c3dd8d76162d.png">

**Comment by ecc521** [comment]
Any IP logging and banning would be server side.

**Comment by FloppyT** [comment]
Ok, and here is where they log your ip

[IMG Link](https://user-images.githubusercontent.com/34033288/82144933-99c85680-9815-11ea-863b-d86f219042a0.png "Script")

**Comment by ecc521** [comment]
No. There is no reason to IP log client side. They know your IP on the server.

**Comment by FloppyT** [comment]
I agree, that is what I meant, that is the section though. They would probably use both Redboat and the request to log you and your ip onto a database.

**Comment by ecc521** [comment]
Whatever they do with IP logging, we can do nothing about it. All we can do is fix the bot so it doesn't get users kicked (which probably triggers bans), and put a bit of pressure on them to remove this.

**Comment by FloppyT** [comment]
There are always ways, but I believe what I've been saying along, bots are going to have to weaker.

**Comment by ecc521** [comment]
The last two actions I have seen from them are going to be painful. For them. Answer rate limiting has kicked out many legitimate players, and this is going to cause absolutely havoc when an entire school gets banned. 

Yes, bots will get weaker. But the only way to remove the advantage of bots is to make everyone the same. And that would destroy the entire game.

**Comment by YodaLightsabr** [comment]
![Um what?](https://user-images.githubusercontent.com/42424091/82160751-88a73600-9865-11ea-89cc-c3dd8d76162d.png) lmfao. They say you're permanently banned from playing games, but there's a button to join a new game.

**Comment by ecc521** [comment]
Yes. Unless something has changed though, you will get blocked from joining a new one.

**Comment by thickman69** [comment]
Ok so you'll not be able to join them any more?? bc I'm trying to get a bot
